packages/editor/src/extensions/key-map/key-map.ts

```diff
@@ -71,7 +86,181 @@ export const KeyMap = Extension.create({
       [tiptapKeys.insertLink.keys]: ({ editor }) => {
         showLinkPopup(editor);
         return true;
+      },
+      [tiptapKeys.moveLineUp.keys]: ({ editor }) => {
+        try {
+          return moveNode(editor, "up", "itself");
+        } catch (e) {
+          console.error("Error moving node up:", e);
+          return false;
+        }
+      },
+      [tiptapKeys.moveLineDown.keys]: ({ editor }) => {
+        try {
+          return moveNode(editor, "down", "itself");
+        } catch (e) {
+          console.error("Error moving node down:", e);
+          return false;
+        }
+      },
+      [tiptapKeys.moveNodeUp.keys]: ({ editor }) => {
+        try {
+          return moveNode(editor, "up", "parent");
+        } catch (e) {
+          console.error("Error moving node up:", e);
+          return false;
+        }
+      },
+      [tiptapKeys.moveNodeDown.keys]: ({ editor }) => {
+        try {
+          return moveNode(editor, "down", "parent");
+        } catch (e) {
+          console.error("Error moving node down:", e);
+          return false;
+        }
       }
     };
   }
 });
+
+function mapChildren<T>(
+  node: Node | Fragment,
+  callback: (child: Node, index: number, frag: Fragment) => T
+): T[] {
+  const array = [];
+  for (let i = 0; i < node.childCount; i++) {
+    array.push(
+      callback(node.child(i), i, node instanceof Fragment ? node : node.content)
+    );
+  }
+  return array;
+}
+
+/**
+ * implementation inspired from https://discuss.prosemirror.net/t/keymap-to-move-a-line/3645/5
+ * @param mode - "itself" | "parent"  - whether to move the current node itself or its immediate valid parent node
+ */
+function moveNode(
+  editor: Editor,
+  dir: "up" | "down",
+  mode: "itself" | "parent"
+) {
```

Comment:
This is a clear indication of a code smell. This function should be refactored into parts:

- moveNodeUp
- moveNodeDown
- moveParentUp
- moveParentDown

If there's anything common between these, you can move it to a separate function. Common does not mean passing conditionals as parameters and using `if` conditions all over the function.